File: question/neetcode/binary_tree/count-good-nodes-in-binary-tree.cpp
#include <bits/stdc++.h>
using namespace std;

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void dfs(TreeNode* node, int max_, int& count){
        if(node->val >= max_) count++;

        max_ = max(max_, node->val);

        if(node->left) dfs(node->left, max_, count);
        if(node->right) dfs(node->right, max_, count);
    }

    int goodNodes(TreeNode* root) {
        int count = 0;

        if(root) dfs(root, -INT_MAX, count);

        return count;
    }
};

File: question/neetcode150/trees/count-good-nodes-in-binary-tree.py
import math

# Definition for a binary tree node.
class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:
    def goodNodes(self, root: TreeNode) -> int:
        ans = 0

        def dfs(node: TreeNode, h):
            nonlocal ans
            if not node:
                return
            
            if node.val >= h:
                ans += 1
            
            dfs(node.left, max(node.val, h))
            dfs(node.right, max(node.val, h))

        dfs(root,-math.inf)
        return ans